### PR TITLE
[ISSUE #7358] Optimize pull request hold service

### LINKED ISSUE #7358: [Enhancement] Low and discontinuous production traffic may cause delays in message consumption
### Before Creating the Enhancement Request

- [X] I have confirmed that this should be classified as an enhancement rather than a bug/feature.

### Summary

https://github.com/apache/rocketmq/discussions/7021

### Motivation

image
image

As the arthas watch results show,  message arriving notify is ~0.1ms ahead of suspending pull request, result in about 5 seconds consumption delay of this message.

### Describe the Solution You'd Like

We can do a check after suspend pull request.

### Describe Alternatives You've Considered

Not yet.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by redlsz**
Fixes #7358

### CODE REVIEW

**Review by joeCarf — APPROVED** [review]
LGTM

File: broker/src/main/java/org/apache/rocketmq/broker/longpolling/PullRequestHoldService.java [hunk 1]
```diff
@@ -54,6 +56,11 @@ public void suspendPullRequest(final String topic, final int queueId, final Pull
 
         pullRequest.getRequestCommand().setSuspended(true);
         mpr.addPullRequest(pullRequest);
+        // check the last notify maxOffset
```
Comment (joeCarf) [inline broker/src/main/java/org/apache/rocketmq/broker/longpolling/PullRequestHoldService.java:59]: if the pull request been put in the `pullRequestTable`, will it be processed repeated?

Comment (redlsz) [reply broker/src/main/java/org/apache/rocketmq/broker/longpolling/PullRequestHoldService.java:59]: Won't. The implementation of notifyMessageArriving guarantees that each request will only be processed once even in concurrent scenarios.

File: broker/src/main/java/org/apache/rocketmq/broker/longpolling/PullRequestHoldService.java [hunk 2]
```diff
@@ -36,6 +36,8 @@ public class PullRequestHoldService extends ServiceThread {
     private final SystemClock systemClock = new SystemClock();
     protected ConcurrentMap<String/* topic@queueId */, ManyPullRequest> pullRequestTable =
         new ConcurrentHashMap<>(1024);
+    protected ConcurrentMap<String/* topic@queueId */, Long/* offset */> lastNotifyMaxOffset =
```
Comment (xdkxlk) [inline broker/src/main/java/org/apache/rocketmq/broker/longpolling/PullRequestHoldService.java:39]: When deleting a topic or reducing the number of queues, the memory of lastNotifyMaxOffset may be leaked.

Comment (redlsz) [reply broker/src/main/java/org/apache/rocketmq/broker/longpolling/PullRequestHoldService.java:39]: > When deleting a topic or reducing the number of queues, the memory of lastNotifyMaxOffset may be leaked.

Indeed. But the leak is very slight? The pullRequestTable has the same issue.

Comment (redlsz) [reply broker/src/main/java/org/apache/rocketmq/broker/longpolling/PullRequestHoldService.java]: > When deleting a topic or reducing the number of queues, the memory of lastNotifyMaxOffset may be leaked.

Resolved. It is more suitable to use expired cache here. @xdkxlk PTAL.

File: broker/src/main/java/org/apache/rocketmq/broker/longpolling/PullRequestHoldService.java [hunk 3]
```diff
@@ -122,6 +134,12 @@ public void notifyMessageArriving(final String topic, final int queueId, final l
     public void notifyMessageArriving(final String topic, final int queueId, final long maxOffset, final Long tagsCode,
         long msgStoreTime, byte[] filterBitMap, Map<String, String> properties) {
         String key = this.buildKey(topic, queueId);
+
+        // record the maxOffset if notify from message store
+        if (msgStoreTime > 0) {
+            this.lastNotifyMaxOffsetCache.put(key, maxOffset);
```
Comment (drpmma) [inline broker/src/main/java/org/apache/rocketmq/broker/longpolling/PullRequestHoldService.java:140]: Should we examine whether the incoming maxOffset exceeds the value of the cache?

Comment (redlsz) [reply broker/src/main/java/org/apache/rocketmq/broker/longpolling/PullRequestHoldService.java:140]: > Should we examine whether the incoming maxOffset exceeds the value of the cache?

You mean the cached offset may be larger than the incoming one? The ReputMessageService executes in the order of commitlog offset, so the notifies would arrive in order of queue logic offset.

**Pull request closed without merging**